> Just for my sanity, can you run a local roachtest that you've changed to t.Fatal (on the main goroutine) and make sure that the test failure will propagate properly?

Sure. I added a `t.Fatalf("testing failure")` to the beginning of the `kv*` roachtests. Running it locally produced:

```
...
--- FAIL: kv0/enc=false/nodes=1 (0.70s)
test artifacts and logs in: artifacts/kv0/enc=false/nodes=1/run_1
        kv.go:57,kv.go:185,test_runner.go:757: testing failure
...
```

Taking the same change and running it on master produced identical looking behavior and output.

--- a/pkg/cmd/roachtest/cluster.go
+++ b/pkg/cmd/roachtest/cluster.go
@@ -2422,44 +2422,27 @@ func (m *monitor) ResetDeaths() {
 	atomic.StoreInt32(&m.expDeaths, 0)
 }
 
-var errGoexit = errors.New("Goexit() was called")
+var errTestFatal = errors.New("t.Fatal() was called")
 
 func (m *monitor) Go(fn func(context.Context) error) {
 	m.g.Go(func() (err error) {
-		var returned bool
 		defer func() {
-			if returned {
-				return
-			}
-			if r := recover(); r != errGoexit && r != nil {
-				// Pass any regular panics through.
-				panic(r)
-			} else {
-				// If the invoked method called runtime.Goexit (such as it
-				// happens when it calls t.Fatal), exit with a sentinel error
-				// here so that the wrapped errgroup cancels itself.
-				//
-				// Note that the trick here is that we panicked explicitly below,
-				// which somehow "overrides" the Goexit which is supposed to be
-				// un-recoverable, but we do need to recover to return an error.
-				err = errGoexit
+			if r := recover(); r != nil {
+				if r != errTestFatal {
+					// Pass any regular panics through.
+					panic(r)
+				}
+				// t.{Skip,Fatal} perform a panic(errTestFatal). If we've caught the
+				// errTestFatal sentinel we transform the panic into an error return so
+				// that the wrapped errgroup cancels itself.
+				err = errTestFatal
 			}
 		}()
 		if impl, ok := m.t.(*test); ok {
 			// Automatically clear the worker status message when the goroutine exits.
 			defer impl.WorkerStatus()
 		}
-		defer func() {
-			if !returned {
-				if r := recover(); r != nil {
-					panic(r)
-				}
-				panic(errGoexit)
-			}
-		}()
-		err = fn(m.ctx)
-		returned = true
-		return err
+		return fn(m.ctx)
 	})
 }
 

--- a/pkg/cmd/roachtest/cluster_test.go
+++ b/pkg/cmd/roachtest/cluster_test.go
@@ -15,7 +15,6 @@ import (
 	"fmt"
 	"os"
 	"regexp"
-	"runtime"
 	"testing"
 	"time"
 
@@ -243,10 +242,9 @@ func TestClusterMonitor(t *testing.T) {
 			// In reality t.Fatal adds text that is returned when the test fails,
 			// so the failing goroutine will be referenced (not like in the expected
 			// error below, where all you see is the other one being canceled).
-			runtime.Goexit()
-			return errors.New("unreachable")
+			panic(errTestFatal)
 		})
-		expectedErr := regexp.QuoteMeta(`Goexit() was called`)
+		expectedErr := regexp.QuoteMeta(`t.Fatal() was called`)
 		if err := m.wait("sleep", "100"); !testutils.IsError(err, expectedErr) {
 			t.Logf("error details: %+v", err)
 			t.Error(err)

--- a/pkg/cmd/roachtest/test.go
+++ b/pkg/cmd/roachtest/test.go
@@ -234,22 +234,22 @@ func (t *test) WorkerProgress(frac float64) {
 	t.progress(goid.Get(), frac)
 }
 
-// Skip records msg into t.spec.Skip and calls runtime.Goexit() - thus
+// Skip records msg into t.spec.Skip and calls panic(errTestFatal) - thus
 // interrupting the running of the test.
 func (t *test) Skip(msg string, details string) {
 	t.spec.Skip = msg
 	t.spec.SkipDetails = details
-	runtime.Goexit()
+	panic(errTestFatal)
 }
 
 // Fatal marks the test as failed, prints the args to t.l, and calls
-// runtime.GoExit(). It can be called multiple times.
+// panic(errTestFatal). It can be called multiple times.
 //
 // If the only argument is an error, it is formatted by "%+v", so it will show
 // stack traces and such.
 //
-// ATTENTION: Since this calls runtime.GoExit(), it should only be called from a
-// test's closure. The test runner itself should never call this.
+// ATTENTION: Since this calls panic(errTestFatal), it should only be called
+// from a test's closure. The test runner itself should never call this.
 func (t *test) Fatal(args ...interface{}) {
 	t.fatalfInner("" /* format */, args...)
 }
@@ -276,7 +276,7 @@ func (t *test) fatalfInner(format string, args ...interface{}) {
 	} else {
 		t.printAndFail(2 /* skip */, args...)
 	}
-	runtime.Goexit()
+	panic(errTestFatal)
 }
 
 // FatalIfErr calls t.Fatal() if err != nil.

--- a/pkg/cmd/roachtest/test_runner.go
+++ b/pkg/cmd/roachtest/test_runner.go
@@ -593,7 +593,9 @@ func (r *testRunner) runTest(
 	defer func() {
 		t.end = timeutil.Now()
 
-		if err := recover(); err != nil {
+		// We only have to record panics if the panic'd value is not the sentinel
+		// produced by t.Fatal*().
+		if err := recover(); err != nil && err != errTestFatal {
 			t.mu.Lock()
 			t.mu.failed = true
 			t.mu.output = append(t.mu.output, t.decorate(0 /* skip */, fmt.Sprint(err))...)
@@ -744,7 +746,9 @@ func (r *testRunner) runTest(
 
 		// This is the call to actually run the test.
 		defer func() {
-			if r := recover(); r != nil {
+			// We only have to record panics if the panic'd value is not the sentinel
+			// produced by t.Fatal*().
+			if r := recover(); r != nil && r != errTestFatal {
 				// TODO(andreimatei): prevent the cluster from being reused.
 				t.Fatalf("test panicked: %v", r)
 			}

